File: topics/type-decl/examples/shape_drawing1.cpp
// =======================
// = Drawing Procedures =
// ======================

// Draw the shapes in the menu
void draw_shapes_menu(shape_type selected_shape)
{
    // Draw the "toolbar" area
    fill_rectangle(ColorLightGrey,0,0,60,600);
    
    // Draw the menu shapes.
    fill_circle(ColorGreen, MENU_CIRCLE);
    fill_rectangle(ColorGreen, MENU_RECT);
    //TODO: Add call to draw Ellipse and Triangle
    
    // Redraw the selected shape
    switch(selected_shape)
    {
        case RECTANGLE: 
            draw_rectangle(ColorBlack, MENU_RECT);
            break;
        case CIRCLE:
            draw_circle(ColorBlack, MENU_CIRCLE);
            break;      //TODO: Add code to draw selected Ellipse and Triangle
        case UNKNOWN:
            break;
    }
}

// Draw a shape onto the screen
void draw_shape(shape &s)
{
    switch(s.type)
    {
        case RECTANGLE:
            fill_rectangle(s.fill_color, s.data.rect);
            break;
        case CIRCLE:
            fill_circle(s.fill_color,s.data.circ);
            break;      //TODO: Add code to draw Ellipse and Triangle shape
        case UNKNOWN:
            break;
    }
}

// Draw the drawing
void draw_drawing(drawing &d)
{
    clear_screen();
    draw_shapes_menu(d.selected_shape);
    
    // Draw all shapes
    for(int i = 0; i< MAX_SHAPES; ++i)
    {
        draw_shape(d.shapes[i]);
    }
}

File: topics/type-decl/examples/ShapeDrawing1.pas
// Draw a shape onto the screen
procedure DrawShape(var s: Shape);
begin
    case s.kind of
        RECTANGLE_TYPE: FillRectangle(s.FillColor, s.rect);
        CIRCLE_TYPE:    FillCircle(s.FillColor, s.circ);
        //Todo: Add Ellipse and Triangle
    end; //end case
end;

procedure DoDrawing(var d: Drawing); // Draw the Drawing
var
    i: integer;
begin   
    // Clear screen and redraw menu
    ClearScreen(ColorWhite);
    DrawShapesMenu(d.SelectedShape);
    
    // Draw the shapes
    for i := 0 to MAX_SHAPES do
        DrawShape(d.Shapes[i]);
end; //do Drawing

// ============================
// = Procedures to Add Shapes =
// ============================

procedure AddRectangle(var s: Shape; pt: Point2d);
begin
    // Set the shape
    s.kind := RECTANGLE_TYPE;
    s.FillColor := ColorRed;
    
    // Copy in the menu Rectangle
    s.rect := MENU_RECT;
    
    // Change its position
    s.rect.x := pt.x;
    s.rect.y := pt.y;
end;

procedure AddCircle(var s: shape; pt: Point2d);
begin
    // Set the shape
    s.kind := CIRCLE_TYPE;
    s.FillColor := ColorBlue;
    
    // Copy in the menu Circle radius
    s.circ.Radius := MENU_CIRCLE.Radius;
    
    // Set the position
    s.circ.Center := pt;
end;//add Circle

//Todo: Add AddEllipse and AddTriangle